### ISSUE TITLE
Exceptions caught and re-thrown in node.js

### ISSUE THREAD

**Issue by w-vi** [kripken/emscripten#4643]
When I am using a emscripten translated C library in JS using node.js and have for example `ReferenceError` in my JS code the exception is caught by the C library and rethrown cluttering the error stack trace.
From the little investigation it seems that it is caused by this bit https://github.com/kripken/emscripten/blob/07b87426f898d6e9c677db291d9088c839197291/src/shell.js#L125

Is there any reason to do it? Is there a way how to avoid it?

Simple example:

The C part `count.c`

``` c
#include <string.h>

int c_count(const char *str) {

    return (int)strlen(str);
}
```

JS preamble `--pre-js`

``` js
var Module = {
  'ready': false,
  'log': function(text) {},
  'logErr': function(text) {},
  'print': function(text) {},
  'printErr': function(text) {},
  'onRuntimeInitialized': function() { this.ready = true; }
};
```

and post `--post-js`

``` js
Module['count'] = function(data, callback) {
  if (false === this.ready) {
    var err = new Error('Module not ready!');
    if (callback) {
      return callback(err, null);
    }
    return err;
  }

  var bufferLen = lengthBytesUTF8(data) + 1;
  var buffer = _malloc(bufferLen);
  stringToUTF8(data, buffer, bufferLen);

  var res = _c_count(buffer);

  if (callback) {
    return callback(null, res);
  }
  return res;
}
```

built using : `emcc -O2 count.c --pre-js pre.js --post-js post.js -o count.js --memory-init-file 0`

And the bad `index.js` code

``` js
const count = require('./count.js');

console.log(xxx); //< here is the ReferenceError
console.log(count.count("123456789"));
```

produces following error when running `node index.js`

``` sh
/home/wvi/src/emcc-crash/count.js:1
aughtException",(function(ex){if(!(ex instanceof ExitStatus)){throw ex}}));Mod
                                                                    ^
ReferenceError: xxx is not defined
    at Object.<anonymous> (/home/wvi/src/emcc-crash/index.js:3:13)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
```

**Comment by w-vi** [comment]
Why I am bothered is that together with node issue https://github.com/nodejs/node/issues/6456 I can't see the original exception as the output gets truncated. Especially if the emscripten built library is big.

**Comment by kripken** [comment]
The reason we do that is the Exit exception is used internally when `exit()` is called. But we don't want to show it as a thrown exception. But sadly that means we need to rethrow.

I guess we could add an option to do do that onexception stuff, but it would mean that in that case, `exit()` wouldn't work properly.

**Comment by w-vi** [comment]
Well in case of using C code as a library the `exit()` will never be called so broken `exit()` call is no issue. Shall I give it a go and make a PR?

**Comment by w-vi** [comment]
Solved by #4689
